src/coordination/coordinator_instance.cpp

```diff
@@ -1144,304 +742,241 @@ auto CoordinatorInstance::AddCoordinatorInstance(CoordinatorToCoordinatorConfig
-  // enabling writing.
-  if (raft_state_->IsCurrentMain(repl_instance.InstanceName())) {
-    spdlog::trace(
-        "Instance {} wasn't considered alive but the health ping passed now. Instance considered current main, trying "
-        "to enable writing.",
-        repl_instance_name);
-    if (!repl_instance.EnableWritingOnMain()) {
-      spdlog::error("Failed to enable writing on instance {}.", repl_instance_name);
-      return;
+  auto lock = std::unique_lock{coord_instance_lock_};
+  auto &instance = FindReplicationInstance(instance_name);
+
+  spdlog::trace("Instance {} performing success callback in thread {}.", instance_name, std::this_thread::get_id());
+
+  instance.OnSuccessPing();
+
+  auto const curr_main_uuid = raft_state_->GetCurrentMainUUID();
+
+  // if I am main and based on Raft log I am the current main I cannot have different UUID and writing is enabled.
+  // Therefore, only the situation where I am replica is handled
+  if (raft_state_->IsCurrentMain(instance_name)) {
+    if (instance_state->is_replica) {
+      auto const is_not_main = [instance_name](auto &&instance) { return instance.InstanceName() != instance_name; };
+      auto repl_clients_info = repl_instances_ | ranges::views::filter(is_not_main) |
+                               ranges::views::transform(&ReplicationInstanceConnector::GetReplicationClientInfo) |
+                               ranges::to<ReplicationClientsInfo>();
+
+      if (!instance.SendPromoteToMainRpc(curr_main_uuid, std::move(repl_clients_info))) {
+        spdlog::error("Failed to promote instance to main with new uuid {}. Trying to do failover again.",
+                      std::string{curr_main_uuid});
+        // running reconcile cluster state in thread pool is not a problem because we have a guarantee that if
+        // TryFailover failed the lock remains opened, hence callbacks won't get executed until lock is closed.
+        // It's possible that the lock didn't even get opened. In that case we will repeat the action.
+        auto const failover_res = TryFailover();
+        switch (failover_res) {
+          case FailoverStatus::SUCCESS: {
+            spdlog::trace("Failover successful after failing to promote main instance.");
+            break;
+          };
+          case FailoverStatus::NO_INSTANCE_ALIVE: {
+            spdlog::trace("Failover failed because no instance is alive.");
+            break;
+          };
+          case FailoverStatus::FAILURE_LOCK_CLOSED: {
+            spdlog::trace("Opening the lock failed during failover.");
+            break;
+          };
+          case FailoverStatus::FAILURE_LOCK_OPENED: {
+            spdlog::trace(
+                "Writing to Raft log partially succeeded during failover, reconciliation task will be scheduled.");
+            thread_pool_.AddTask([this]() { this->ReconcileClusterState(); });
```

Comment:
I would put this in a function TryReconcile where we first set the ready to false and then put it in the queue.

Reply:
makes sense

Reply:
will be changed when I remove opening the lock and is_leader_ready. Now this is protected with IsLockOpened